### PR TITLE
add pageExtensions support to appDir

### PR DESCRIPTION
**Pull request by ForsakenHarmony**
The previous PR #3499 only implemented support for `pages` directory.

The `app` directory previously had no validation at all, it's added with this PR.

### CODE REVIEW

**Comment by ForsakenHarmony** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3510** <a href="https://app.graphite.dev/github/pr/vercel/turbo/3510" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/turbo/3510?utm_source=stack-comment).

File: crates/next-core/src/app_source.rs [hunk 1]
```diff
@@ -344,131 +349,149 @@ async fn create_app_source_for_directory(
...
+    let DirectoryContent::Entries(entries) = &*input_dir.read_dir().await? else {
+        return Ok(CombinedContentSource { sources }.cell());
+    };
+
+    let allowed_extensions = &*page_extensions.await?;
+
+    for (name, entry) in entries.iter() {
+        if let &DirectoryEntry::File(file) = entry {
+            if let Some((name, ext)) = name.rsplit_once('.') {
+                if !allowed_extensions.iter().any(|allowed| allowed == ext) {
```
Comment (alexkirsz) [inline crates/next-core/src/app_source.rs:362]: Should this perhaps only apply to `page`/`error`/`loading`?

Comment (ForsakenHarmony) [reply crates/next-core/src/app_source.rs:362]: I don't think so, why?

Comment (alexkirsz) [reply crates/next-core/src/app_source.rs:362]: It feels weird that `pageExtensions` would apply to things that aren't technically pages, like `head` or `layout`, but I guess it makes sense to support these too, so 🤷

Comment (ForsakenHarmony) [reply crates/next-core/src/app_source.rs:362]: Well, we're just copying next.js behaviour

It treats all of them the same as far as I can tell https://github.com/vercel/next.js/blob/canary/packages/next/src/build/webpack/loaders/next-app-loader.ts#L12-L12

Comment (jridgewell) [inline crates/next-core/src/app_source.rs:362]: Should we be representing this with a `BSet` instead?

**Review by jridgewell — APPROVED** [review]

**Review by sokra — APPROVED** [review]

**Pull request merged**